File: typescript/2024/day01.ts
const aoc = require("./../aoc");

interface AocInput {
    text: string;
    lines: string[];
}

function zip<T>(arr1: T[], arr2: T[]) {
    return arr1.map((element, index) => [element, arr2[index]]);
}

function instructions(lines: string[], col: number): number[] {
    return lines.map(line => parseInt(line.split("   ")[col])).toSorted();
}

function part1(input: AocInput): number {
    const left = instructions(input.lines, 0);
    const right = instructions(input.lines, 1);
    return zip(left, right).reduce((sum, [l, r]) => sum + Math.abs(l - r), 0);
}

function part2(input: AocInput): number {
    return 0;
}

async function main() {
    const test = await aoc.findAocInput("./2024/data/Day01_Test.txt");
    const real = await aoc.findAocInput("./2024/data/Day01.txt");
    console.log("Part 1:", part1(test), part1(real), "expected", 1660292);
    console.log("Part 2:", part2(test), part2(real), "expected", 22776016);
}

main().catch(err => console.error(err));

File: golang/2024/day01.go
package day01

import (
	"aoc"
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
)

func Part1(input *aoc.AocInput) int {
	lines := input.Lines
	left := instructions(lines, 0)
	right := instructions(lines, 1)

	sum := 0
	for i := 0; i < len(left); i++ {
		sum += int(math.Abs(float64(left[i] - right[i])))
	}

	return sum
}

func Part2(input *aoc.AocInput) int {
	lines := input.Lines
	left := instructions(lines, 0)
	right := instructions(lines, 1)
	counts := countOccurrences(right)

	sum := 0
	for _, id := range left {
		count := counts[id]
		sum += count * id
	}

	return sum
}

func instructions(lines []string, col int) []int {
	var result []int
	for _, line := range lines {
		nums := strings.Fields(line)
		if len(nums) > col {
			result = append(result, parseInt(nums[col]))
		}
	}

	sort.Ints(result)
	return result
}

func countOccurrences(nums []int) map[int]int {
	counts := make(map[int]int)
	for _, num := range nums {
		counts[num]++
	}

	return counts
}

func parseInt(s string) int {
	val, err := strconv.Atoi(s)
	if err != nil {
		fmt.Println("Error converting string to int:", err)
	}
	return val
}
